src/Npgsql/NpgsqlDataReader.cs

```diff
@@ -52,6 +53,7 @@ public sealed class NpgsqlDataReader : DbDataReader, IDbColumnSchemaGenerator
     internal ReaderState State = ReaderState.Disposed;
 
     internal NpgsqlReadBuffer Buffer = default!;
```

Comment:
I'm not sure, I haven't touched this line.

Reply:
Reason being, you added a `ReferenceEqual` check to `NpgsqlDataReader` to replace the buffer if it differs from the one connector has. My thoughts were that it might pgo a bit to assume that the buffer almost never changes.

Reply:
I added the check so we avoid the write barrier of unconditionally setting the buffer reference. Afaik PGO cannot assume anything that isn't a static readonly will stay constant (or do anything fancy to optimize a reference check).

Reply:
> I added the check so we avoid the write barrier of unconditionally setting the buffer reference. Afaik PGO cannot assume anything that isn't a static readonly will stay constant (or do anything fancy to optimize a reference check).

Yeah, currently it can reason about the implementation of the interface (testing first again the most probable), don't think there is anything what I described yet (doesn't mean it will not be implemented at some point).
Anyway, that's more of a suggestion (we don't really lose anything from it) and I'm okay not doing it.